packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -1795,24 +1781,32 @@ class OpacityLayer extends ContainerLayer {
       return true;
     }());
 
-    if (enabled)
+    final int realizedAlpha = alpha!;
+    // The type assertions work because the [alpha] setter nulls out the
+    // engineLayer if it would have changed type (i.e. changed to or from 255).
+    if (enabled && realizedAlpha < 255) {
+      assert(_engineLayer is ui.OpacityEngineLayer?);
       engineLayer = builder.pushOpacity(
-        alpha!,
-        offset: offset! + layerOffset,
+        realizedAlpha,
+        offset: offset + layerOffset,
         oldLayer: _engineLayer as ui.OpacityEngineLayer?,
       );
-    else
-      engineLayer = null;
+    } else {
+      assert(_engineLayer is ui.OffsetEngineLayer?);
+      engineLayer = builder.pushOffset(
```

Comment:
What is the value of subclassing OffsetLayer if we push our own?

Reply:
Eliminates duplicated code with regard to setting the transform, and both layers provide an `offset` property.

Users of this API don't really deal directly with scene builder - but in some of my explorations around trying to optimize layer usage, it can be helpful to know whether a layer has an offset or not.